File: crates/bridge_rpc/src/bridge/builder.rs
use std::{
    collections::{HashMap, HashSet},
    fmt::Display,
};

use crate::{
    BridgeRpcBuilderError, BridgeRpcBuilderErrorInner, RequestContext,
    StreamContext, StreamError, StreamHandlerFn, Transport,
    bridge::{self, BridgeRpc, RequestHandlerFn},
};

pub struct BridgeRpcBuilder<TTransport: Transport> {
    transport: TTransport,
    request_handlers: HashMap<String, RequestHandlerFn>,
    stream_handlers: HashMap<String, StreamHandlerFn>,
}

impl<TTransport: Transport> BridgeRpcBuilder<TTransport> {
    pub fn new(transport: TTransport) -> Self {
        Self {
            transport,
            request_handlers: HashMap::new(),
            stream_handlers: HashMap::new(),
        }
    }

    pub fn request_handler<TRequestData, TResponseData, TError, TFuture, TFn>(
        mut self,
        path: impl Into<String>,
        handler: TFn,
    ) -> Self
    where
        TFn: Fn(RequestContext<TRequestData>) -> TFuture
            + Send
            + Sync
            + Clone
            + 'static,
        TRequestData: for<'de> serde::Deserialize<'de>,
        TResponseData: serde::Serialize,
        TError: Display,
        TFuture:
            Future<Output = Result<TResponseData, TError>> + Send + 'static,
    {
        self.request_handlers
            .insert(path.into(), bridge::create_request_handler(handler));
        self
    }

    pub fn stream_handler<TStartData, TStreamData, TError, TFuture, TFn>(
        mut self,
        path: impl Into<String>,
        handler: TFn,
    ) -> Self
    where
        TStartData: for<'de> serde::Deserialize<'de>,
        TStreamData: for<'de> serde::Deserialize<'de>,
        TError: Display,
        TFuture: Future<Output = Result<(), TError>> + Send + 'static,
        TFn: Fn(StreamContext<TStartData, TStreamData, StreamError>) -> TFuture
            + Send
            + Sync
            + Clone
            + 'static,
    {
        self.stream_handlers
            .insert(path.into(), bridge::create_stream_handler(handler));
        self
    }

    fn validate_paths(&self) -> Result<(), BridgeRpcBuilderError> {
        let paths = self
            .request_handlers
            .keys()
            .chain(self.stream_handlers.keys())
            .collect::<HashSet<_>>();

        for path in paths {
            if self.stream_handlers.contains_key(path)
                && self.request_handlers.contains_key(path)
            {
                return Err(BridgeRpcBuilderErrorInner::DuplicatePath(
                    path.clone(),
                )
                .into());
            }
        }

        Ok(())
    }

    pub fn build(self) -> Result<BridgeRpc<TTransport>, BridgeRpcBuilderError> {
        self.validate_paths()?;

        Ok(BridgeRpc::new(
            self.transport,
            self.request_handlers,
            self.stream_handlers,
        ))
    }
}

#[cfg(test)]
mod tests {
    use tokio_stream::StreamExt as _;

    use crate::MockTransport;

    use super::*;

    #[tokio::test]
    async fn test_builder_request_handler() {
        let transport = MockTransport::new();
        let rpc = BridgeRpcBuilder::new(transport)
            .request_handler(
                "test_path",
                |req: RequestContext<String>| async move {
                    Ok::<_, String>(req.data)
                },
            )
            .build()
            .expect("should be able to build");

        assert!(
            rpc.has_request_handler("test_path").await,
            "handler should be registered"
        );
    }

    #[tokio::test]
    async fn test_builder_stream_handler() {
        let transport = MockTransport::new();
        let rpc = BridgeRpcBuilder::new(transport)
            .stream_handler("test_path", |mut s: StreamContext| async move {
                while let Some(_) = s.stream.next().await {}

                Ok::<_, String>(())
            })
            .build()
            .expect("should be able to build");

        assert!(
            rpc.has_stream_handler("test_path").await,
            "handler should be registered"
        );
    }
}

File: packages/bridge-rpc/src/bridge/builder.ts
import type { Transport } from "../transport";
import {
    type BridgeRequestHandler,
    BridgeRpc,
    type BridgeStreamHandler,
    type UnknownBridgeRequestHandler,
    type UnknownBridgeStreamHandler,
} from "./bridge-impl";
import type { Push } from "./type-utils";

export class DuplicatePathError extends Error {
    constructor(public readonly path: string) {
        super(
            `Duplicate path is registered as a stream and request handler: "${path}", path must be unique`,
        );
    }
}

export class BridgeRpcBuilder<
    TRequestHandlers extends [...string[]] = [],
    TStreamHandlers extends [...string[]] = [],
> {
    private requestHandlers = new Map<string, UnknownBridgeRequestHandler>();
    private streamHandlers = new Map<string, UnknownBridgeStreamHandler>();

    static create(transport: Transport) {
        return new BridgeRpcBuilder(transport);
    }

    constructor(private readonly transport: Transport) {}

    requestHandler<const TPath extends string, TRequest, TResponse>(
        path: TPath,
        handler: BridgeRequestHandler<TRequest, TResponse>,
    ) {
        this.requestHandlers.set(path, handler as UnknownBridgeRequestHandler);

        return this as unknown as BridgeRpcBuilder<
            Push<TRequestHandlers, TPath>,
            TStreamHandlers
        >;
    }

    streamHandler<const TPath extends string, TStartData, TStreamData>(
        path: TPath,
        handler: BridgeStreamHandler<TStartData, TStreamData>,
    ) {
        this.streamHandlers.set(path, handler as UnknownBridgeStreamHandler);

        return this as unknown as BridgeRpcBuilder<
            TRequestHandlers,
            Push<TStreamHandlers, TPath>
        >;
    }

    build(): BridgeRpc<TRequestHandlers, TStreamHandlers> {
        const duplicate = getDuplicatePath(
            this.requestHandlers,
            this.streamHandlers,
        );
        if (duplicate !== null) {
            throw new DuplicatePathError(duplicate);
        }

        return new BridgeRpc({
            transport: this.transport,
            requestHandlers: this.requestHandlers,
            streamHandlers: this.streamHandlers,
        });
    }
}

function getDuplicatePath(
    map1: Map<string, unknown>,
    map2: Map<string, unknown>,
): string | null {
    const keys = new Set([...map1.keys(), ...map2.keys()]);

    for (const key of keys) {
        if (map1.has(key) && map2.has(key)) {
            return key;
        }
    }
    return null;
}
